File: backend/src/seed/seed.ts
import mongoose from "mongoose";
import dotenv from "dotenv";
import Author from "../models/Author";
import Book from "../models/Book";
import Review from "../models/Review";
import Sale from "../models/Sale";
import { faker } from "@faker-js/faker";

dotenv.config();

async function seed(): Promise<void> {
  try {
    await mongoose.connect(process.env.MONGO_URI || "");

    console.log("⚡ Clearing old data...");
    await Author.deleteMany({});
    await Book.deleteMany({});
    await Review.deleteMany({});
    await Sale.deleteMany({});

    console.log("⚡ Seeding Authors...");
    const authors = [];
    for (let i = 0; i < 50; i++) {
      authors.push(
        new Author({
          name: faker.person.fullName(),
          date_of_birth: faker.date.birthdate({
            min: 1950,
            max: 1995,
            mode: "year",
          }),
          country: faker.location.country(),
          description: faker.lorem.sentence(),
        }),
      );
    }
    const savedAuthors = await Author.insertMany(authors);

    console.log("⚡ Seeding Books + Reviews + Sales...");
    for (let i = 0; i < 300; i++) {
      const author = faker.helpers.arrayElement(savedAuthors);

      const book = await Book.create({
        name: faker.lorem.words(3),
        summary: faker.lorem.paragraph(),
        publication_date: faker.date.past({ years: 30 }),
        author: author._id,
        total_sales: 0, // will update later
      });

      // Reviews (1–10 per book)
      const reviews = [];
      const numReviews = faker.number.int({ min: 1, max: 10 });
      for (let j = 0; j < numReviews; j++) {
        reviews.push({
          book: book._id,
          review: faker.lorem.sentence(),
          score: faker.number.int({ min: 1, max: 5 }),
          up_votes: faker.number.int({ min: 0, max: 5000 }),
        });
      }
      await Review.insertMany(reviews);

      // Sales (5 years)
      let totalSales = 0;
      const currentYear = new Date().getFullYear();
      for (let y = 0; y < 5; y++) {
        const yearSales = faker.number.int({ min: 1000, max: 100000 });
        totalSales += yearSales;
        await Sale.create({
          book: book._id,
          year: currentYear - y,
          sales: yearSales,
        });
      }

      // Update total_sales on book
      book.total_sales = totalSales;
      await book.save();
    }

    console.log("✅ Seeding complete!");
    mongoose.connection.close();
  } catch (err) {
    console.error("❌ Seeding error:", err);
    process.exit(1);
  }
}

seed();

File: core/management/commands/seed.py
import random
from datetime import date
from django.core.management.base import BaseCommand
from faker import Faker
from mongoengine.connection import get_db
from dotenv import load_dotenv
import os

from core.models import Author, Book, Review, Sale
from core.search import (
    bulk_index_authors,
    bulk_index_books,
    bulk_index_reviews,
    bulk_index_sales,
)

faker = Faker()
load_dotenv()
ES_ENABLED = os.getenv("ES_ENABLED", "false").lower() == "true"


class Command(BaseCommand):
    help = "Seed MongoDB with authors, books, reviews, and sales."

    def handle(self, *args, **kwargs):
        db = get_db()

        self.stdout.write("⚡ Dropping old collections...")
        db.drop_collection("author")
        db.drop_collection("book")
        db.drop_collection("review")
        db.drop_collection("sale")

        self.stdout.write("⚡ Seeding Authors...")
        authors = []
        for _ in range(50):
            author = Author(
                name=faker.name(),
                birthday=faker.date_of_birth(minimum_age=30, maximum_age=75),
                origin_country=faker.country(),
                description=faker.sentence(),
            )
            author.save()
            authors.append(author)

        self.stdout.write("⚡ Seeding Books + Reviews + Sales...")
        for _ in range(300):
            author = random.choice(authors)

            book = Book(
                name=" ".join(faker.words(3)),
                summary=faker.paragraph(),
                publication_date=faker.date_between(
                    start_date="-30y", end_date="today"
                ),
                author=author,
            )
            book.save()

            # Reviews (1–10 per book)
            for _ in range(random.randint(1, 10)):
                Review(
                    book=book,
                    score=random.randint(1, 5),
                    up_votes=random.randint(0, 5000),
                ).save()

            # Sales (last 5 years)
            total_sales = 0
            current_year = date.today().year
            for y in range(5):
                year_sales = random.randint(1000, 100000)
                total_sales += year_sales
                Sale(book=book, year=current_year - y, count=year_sales).save()

            book.total_sales = total_sales
            book.save()

        print("Is Elasticsearch enabled?", ES_ENABLED)
        if not ES_ENABLED:
            self.stdout.write(self.style.SUCCESS("✅ Seeding complete!"))
            return

        self.stdout.write(self.style.SUCCESS("✅ Seeding complete!"))
        self.stdout.write("⚡ Indexing Books in Elasticsearch...")
        bulk_index_books(Book.objects)
        self.stdout.write("⚡ Indexing Authors in Elasticsearch...")
        bulk_index_authors(Author.objects)
        self.stdout.write("⚡ Indexing Reviews in Elasticsearch...")
        bulk_index_reviews(Review.objects)
        self.stdout.write("⚡ Indexing Sales in Elasticsearch...")
        bulk_index_sales(Sale.objects)
        self.stdout.write(self.style.SUCCESS("✅ Indexing complete!"))
